src/shared/components/dataAccessTokens/DataAccessTokens.tsx

```diff
@@ -0,0 +1,102 @@
+import * as React from "react";
+import AppConfig from "appConfig";
+import client from "shared/api/cbioportalClientInstance";
+import {AppStore} from "../../../AppStore";
+import LoadingIndicator from "../loadingIndicator/LoadingIndicator";
+import {observer} from "mobx-react";
+import {observable} from "mobx";
+import {buildCBioPortalPageUrl} from "../../api/urls";
+import {isNullOrUndefined} from "util";
+import fileDownload from 'react-file-download';
+import classnames from "classnames";
+
+export class UserDataAccessToken {
+    @observable token : string;
+    @observable creationDate: string;
+    @observable expirationDate: string;
+    @observable username: string;
+    constructor(token: string, creationDate: string, expirationDate: string, username: string) {
+        this.token = token;
+        this.creationDate = creationDate;
+        this.expirationDate = expirationDate;
+        this.username = username;
+    }
+}
+
+export interface IDataAccessTokensProps {
+    token?:string;
+    creationDate?:string;
+    expirationDate?:string;
+    loadingComponent?:JSX.Element;
+    appStore:AppStore;
+}
+
+export interface IDataAccessTokensState {
+    userDataAccessToken:UserDataAccessToken|undefined;
+}
+
+function buildDataAccessTokenFileContents(dat:UserDataAccessToken | undefined) {
+    if (!isNullOrUndefined(dat)) {
+        const fileContents = "token: " + dat!.token + "\n" +
+            "creation_date: " + dat!.creationDate + "\n" +
+            "expiration_date: " + dat!.expirationDate;
+        return fileContents;
+    } else {
+        alert("Cannot create Data Access Token file for user with non-existent tokens.");
+        return null;
+    }
+}
+
+function downloadDataAccessTokenFile(dat:UserDataAccessToken | undefined) {
+    if (!isNullOrUndefined(dat)) {
+        const fileContents = buildDataAccessTokenFileContents(dat);
+        fileDownload(fileContents, "cbioportal_data_access_token.txt");
+    }
+}
+
+@observer
+export class DataAccessTokens extends React.Component<IDataAccessTokensProps, IDataAccessTokensState> {
+    public static defaultProps: Partial<IDataAccessTokensProps> = {
+        loadingComponent:<LoadingIndicator isLoading={true}/>
+    };
+
+    constructor(props: IDataAccessTokensProps) {
+        super(props);
+        this.state = {
+            userDataAccessToken: new UserDataAccessToken('','','','')
+        } as IDataAccessTokensState;
+    }
+
+     async generateNewDataAccessToken() {
+        if (this.props.appStore.isLoggedIn) {
+            let _token = await Promise.resolve(
+                client.createDataAccessTokenUsingPOST(
+                    {'allowRevocationOfOtherTokens':AppConfig.serverConfig.dat_revoke_other_tokens}))
+            const dat = new UserDataAccessToken(_token.token, _token.creation, _token.expiration, _token.username);
```

Comment:
I'm not able to work out in my mind whether this code will be executed before the user clicks the "Download Token" menu choice. It is important that the frontend *not* call the API and generate the token before the user chooses the menu choice, because every time a token is generated, the previous token will be automatically revoked and this would revoke a user's current token just for signing in to the portal. Also, it would be undesirable to generate a token for every user who logs into the portal -- most likely very few users will ever request a token. So I think it is important for us to test whether an existing token for a user gets revoked when that user simply logs in to the portal again with a different browser (i.e. with no cookie and where they authenticate)
My guess is that when the component (DataAccessTokens.tsx) becomes part of the DOM, it will call generateNewDataAccessToken(), and that will instantiate UserDataAccessToken with the *resolved* promise that yields _token. So I think the token will be generated when the DOM element is realized ... perhaps immediately ... or perhaps it becomes part of the DOM after triggering the dropdown menu? (So when testing revocation of prior tokens in an incognito window, maybe we should drop the menu down but not choose "Download Token" ... and then see if the revocation occurred)

Reply:
This has been resolved by directly calling the download function in the click handler for the menu item .. rather than in a separate data access token component / observer / store.